### PR TITLE
Patch issue 21 (Issues on OpenWRT 22.03.4)

### LINKED ISSUE #21: Issues on OpenWRT 22.03.4
Hi,

thank you for your tremendous work. This was, what I was looking for.

Thanks to #18 by @LLdaniel I was also able to install it on OpenWrt 22.03.4.

Unfortunately, I found some issues:

1. Missing r-permission on config
/etc/init.d/hetzner_ddns starts process as nobody. After installation, the config file has o600 permissions only
2. Missing write-permissions on log-file and pid-file
/etc/init.d/hetzner_ddns starts process as nobody. /var/run and /var/log are by default only root-writable
3. Wrong link to start script
/etc/init.d/hetzner_ddns wants to start /usr/bin/hetzner_ddns, but only /usr/bin/hetzner_ddns.sh exists
4. Issue with config-file as parameter
After solving all issues above, the following issue remains:  The start-up script tries to acquire /var/log/hetzner_ddns./etc/config/hetzner_ddns.conf.log

My current workarounds to this issue:

1. After installation: chmod a+r /etc/config/hetzner_dds.conf
2. After installation: touch /var/run/hetzner_ddns.pid /var/log/hetzner_ddns.pid && chown nobody:nogroup /var/run/hetzner_ddns.pid /var/log/hetzner_ddns.pid
3. After installation: ln -s /usr/bin/hetzner_ddns.sh /usr/bin/hetzner_ddns
4. After installation: sed -i 's@command /usr/bin/hetzner_ddns -d /etc/config/hetzner_ddns.conf@command /usr/bin/hetzner_ddns -d@' /etc/init.d/hetzner_ddns

Step four is something I really did not like. And I think, that 1-3 should be trivial steps I can also provide a fix for. 

But I have no idea what goes wrong in 4.

Regards

Carsten

### PR DESCRIPTION
**Pull request by carstencodes**
- Rename hetzner_ddns.sh to hetzner_ddns
- Add read permission to config file
- Remove config-file from startup-script
- Set correct permissions on /var/run and /var/log related files

Closes #21 

/cc @LLdaniel

### CODE REVIEW

File: release/OpenWrt/Makefile [hunk 1]
```diff
@@ -59,13 +59,13 @@ define Package/$(PKG_NAME)/install
...
 	$(INSTALL_DIR) $(1)/usr/bin
-	$(INSTALL_BIN) ./files/hetzner_ddns.sh $(1)/usr/bin
+	$(INSTALL_BIN) ./files/hetzner_ddns $(1)/usr/bin
 
 	$(INSTALL_DIR) $(1)/etc/init.d
 	$(INSTALL_BIN) ./files/hetzner_ddns.openwrt.rc $(1)/etc/init.d/hetzner_ddns
 
 	$(INSTALL_DIR) $(1)/etc/config
-	$(INSTALL_CONF) ./files/hetzner_ddns.conf $(1)/etc/config/hetzner_ddns.conf
+	$(INSTALL_DATA) ./files/hetzner_ddns.conf $(1)/etc/config/hetzner_ddns.conf
```
Comment (LLdaniel) [inline release/OpenWrt/Makefile:68]: Ah, is it better to do it with `$(INSTALL_DATA)` vs. `$(INSTALL_CONF)`?

File: hetzner_ddns.openwrt.rc [hunk 2]
```diff
@@ -4,9 +4,24 @@
 START=99
 USE_PROCD=1
 
+_LOG_FILE=/var/log/hetzner_ddns.log
+_PID_FILE=/var/run/hetzner_ddns.pid
+
+fix_permissions() {
```
Comment (LLdaniel) [inline hetzner_ddns.openwrt.rc:10]: Should we also set the owner of the config file to `nobody:nogroup`?

I did encounter issues when leaving it to `root:root`. 
What I did:
- changed the config file contents in order to match my zone on hetzner
- restarted the service with `service hetzner_ddns restart`
- had a look in my log file which showed me the following:
```
[2024-12-27 19:52:00] Started Hetzner DDNS daemon
[2024-12-27 19:52:00] Error: unable to read configuration file /etc/config/hetzner_ddns.conf
```
- after changing the owner of `hetzner_ddns.conf` to `nobody:nogroup` I could process the config file, because the log showed me this:
```
[2024-12-27 20:26:07] Started Hetzner DDNS daemon
[2024-12-27 20:26:07] Reading configuration from /etc/config/hetzner_ddns.conf
[2024-12-27 20:26:07] Info: update of A records not set, enabling by default
[2024-12-27 20:26:07] Info: update of AAAA records not set, enabling by default
[2024-12-27 20:26:09] Zone for refractedlight.eu: <redacted>
[2024-12-27 20:26:09] IPv4 record for cloud.refractedlight.eu: <redacted>
[2024-12-27 20:26:09] IPv6 record for cloud.refractedlight.eu: (missing)
[2024-12-27 20:26:09] Configuration successful
[2024-12-27 20:26:09] Watching for IP address and record changes
```
- then I also could confirm that the process is running:
```
 # ps | grep hetzner
17873 nobody    1564 S    {hetzner_ddns} /bin/sh /usr/bin/hetzner_ddns -d
```

But maybe this was just something strange on my installation... I am not that sure. 
Did test that with 
[hetzner-ddns_0.2.7-1_all.ipk.txt](https://github.com/user-attachments/files/18263052/hetzner-ddns_0.2.7-1_all.ipk.txt)
 (`.txt` suffix only because of GitHub rejecting pure `.ipk`)

**Review by LLdaniel — COMMENTED** [review]
Hi @carstencodes, 
thank you for your patch! Appreciate it!

I felt so free and commented two small things and also reported from my test. If you want, please feel free to look into the review details.

Despite the strange behaviour around the owner of the config file (and the two comments) this looks good to me. 

I also was able to spawn a manual daemon process as root, without managing the `hetzner_ddns` with the `service` command by simply typing: `hetzner_ddns -d`. So this still works in addition, if someone don't want to use the init script.

Also permissions on `.pid` and log `.log` look good:
```
 # ll /var/log/hetzner_ddns.log 
-rw-r--r--    1 nobody   nogroup       8570 Dec 27 20:26 /var/log/hetzner_ddns.log

 # ll /var/run/hetzner_ddns.pid 
-rw-r--r--    1 nobody   nogroup          6 Dec 27 20:26 /var/run/hetzner_ddns.pid
```

I also tested a change for the IP address, which worked fine (currently I have no IPv6 entry):
```
[2024-12-27 20:44:20] Update IPv4 for cloud.refractedlight.eu: 172.0.0.1 => 84.135.98.123
```

File: release/OpenWrt/Makefile (hunk 1 above)
Comment (carstencodes) [reply release/OpenWrt/Makefile:68]: Yes, as mentioned in https://github.com/filiparag/hetzner_ddns/pull/22/commits/ec11febc74acc63381d3aa7ad6d61149eaf5450e.

According to https://openwrt.org/docs/guide-developer/packages#file_installation_macros 

INSTALL_CONF is short for install -m600

while

INSTALL_DATA is short for install -m644

This should give read access to all users

File: release/OpenWrt/Makefile [hunk 3]
```diff
@@ -50,7 +50,7 @@ endef
 define Package/$(PKG_NAME)/install
 	# copy shell script, init.d file and conf as preparation
 	mkdir -p files
-	cp $(PKG_BUILD_DIR)/hetzner_ddns.sh ./files/
+	cp $(PKG_BUILD_DIR)/hetzner_ddns.sh ./files/hetzner_ddns
```
[the remark this answers is not in the document]
Comment (carstencodes) [reply release/OpenWrt/Makefile:53]: Done in https://github.com/carstencodes/hetzner_ddns/commit/79d57d66f3b7f1c68eb7a3f9474f83093e9621b0

File: hetzner_ddns.openwrt.rc (hunk 2 above)
Comment (carstencodes) [reply hetzner_ddns.openwrt.rc:10]: This is strange. I will give it some further investigations in the course of the weekend, but what permissions did your config file have after installation/editing?

Comment (LLdaniel) [reply hetzner_ddns.openwrt.rc:10]: Sorry for the fuss, seems that I did something wrong previously or had some older configuration files.
<details><summary>Installation history</summary>
<code>
root@danielcpe ~
 # opkg install hetzner-ddns_0.2.7-1_all.ipk 
Installing hetzner-ddns (0.2.7-1) to root...
Configuring hetzner-ddns.
</code>

<code>
root@danielcpe /etc/config
 # ll hetzner_ddns.conf
-rw-r--r--    1 root     root           453 Dec 27 14:34 hetzner_ddns.conf
</code>
<code>
root@danielcpe /etc/config
 # cat hetzner_ddns.conf
interval=60                             # Seconds between updates / TTL value
key='********************************'  # Hetzner DNS API key
domain='example.com'                    # Top level domain name
records='homelab media vpn'             # Space separated host subdomains (@ for domain itself)
ipv4=true                               # Enable updating A records (IPv4)
ipv6=true                               # Enable updating AAAA records (IPv6)
</code>
<code>
root@danielcpe /etc/config
 # cat /var/log/hetzner_ddns.log 
[2024-12-29 17:14:55] Started Hetzner DDNS daemon
[2024-12-29 17:14:55] Reading configuration from /etc/config/hetzner_ddns.conf
[2024-12-29 17:14:56] Error: Invalid API key
</code>
</details> 

Now the configuration file is owned by root, but the daemon process is running and owned by nobody and the update works: 
```
16841 nobody    1564 S    {hetzner_ddns} /bin/sh /usr/bin/hetzner_ddns -d
```
```
[2024-12-29 17:32:34] Watching for IP address and record changes
[2024-12-29 17:36:14] Update IPv4 for cloud.refractedlight.eu: 127.0.0.1 => 84.135.98.106
```

TL;DR Seems fine now :heavy_check_mark:

**Review by filiparag — APPROVED** [review]

**Comment by carstencodes** [comment]
Sorry for coming back to you so late with this - I was in a terrible health condition.

If you both agree, we can merge this PR.

Regards

Carsten

**Comment by carstencodes** [comment]
@filiparag Could you please merge this PR?
